### PR TITLE
Typeorm updation

### PR DESCRIPTION
**Pull request by RidhamShah**
This PR covers below changes:

1. Update typeORM and all related depreciated code to new code.
2. Now, the export replica hostname would also be an array as a separate env variable `TYPEORM_HOSTNAME_EXPORT_REPLICA` and rest replicas (if any) would be in the `TYPEORM_HOSTNAME_REPLICAS`. For now you need to move the value in `TYPEORM_HOSTNAME_REPLICAS` to `TYPEORM_HOSTNAME_EXPORT_REPLICA`.
3. We updated the `typeormLoader` file to have the mainDB connection use the 2nd read replica as slaves which will be automatically used for any read operation that uses the main DB connection by default. Only the export queries would use the export connection which will hit the 1st read replica. In case we dont have a 2nd read replica, all the reads apart from export queries would hit the main RDS instance.

### CODE REVIEW

File: backend/packages/Upgrade/src/loaders/typeormLoader.ts [hunk 1]
```diff
@@ -67,8 +199,74 @@ export const typeormLoader: MicroframeworkLoader = async (settings: Microframewo
     synchronize: env.db.synchronize,
     logging: env.db.logging as boolean | 'all' | LogLevel[],
     maxQueryExecutionTime: env.db.maxQueryExecutionTime,
-    entities: env.app.dirs.entities,
-    migrations: env.app.dirs.migrations,
+    entities: [
```
Comment (ppratikcr7) [inline backend/packages/Upgrade/src/loaders/typeormLoader.ts:202]: @RidhamShah Can you make a common array outside and use in both master and export connection, as its redundant.

Comment (RidhamShah) [reply backend/packages/Upgrade/src/loaders/typeormLoader.ts]: code updated

**Comment by danoswaltCL** [comment]
Currently all of our CL client lib usage is on 5.0, which is fine right now, 5.1 is not required. When this code gets promoted, since still synced with `dev`, it will carry everything currently in dev also as 5.2. 

5.2 shouldn't be required either. Are we sure that this 5.2 TypeOrm update version won't break compatibility with CL apps using 5.0? I ask because a change like [this](https://github.com/CarnegieLearningWeb/UpGrade/pull/1692) should not be merged in yet, that would break all of our clients on staging and we would need to get them to update their libraries / usage. Can we maybe cease pulling dev and "freeze" into branch this so we don't pull in accidental surprises? Then we can pip dev to like 5.3? @VivekFitkariwala @RidhamShah @bcb37

File: backend/packages/Upgrade/src/api/controllers/ExperimentClientController.ts [hunk 2]
```diff
@@ -1014,7 +1014,7 @@ export class ExperimentClientController {
   @Delete('clearDB')
   public async clearDB(@Req() request: AppRequest): Promise<string> {
     // if DEMO mode is enabled, then clear the database:
-    if (env.app.demo) {
+    if (!env.app.demo) {
```
Comment (danoswaltCL) [inline backend/packages/Upgrade/src/api/controllers/ExperimentClientController.ts:1017]: why this change?

File: backend/packages/Upgrade/.env.example [hunk 3]
```diff
@@ -22,6 +22,7 @@ LOG_OUTPUT=dev
 #
 TYPEORM_CONNECTION=postgres
 TYPEORM_HOST=localhost
+TYPEORM_HOSTNAME_EXPORT_REPLICA=[]
```
Comment (danoswaltCL) [inline backend/packages/Upgrade/.env.example:25]: is the separate new env var needed for the typeorm update? does it need to be an array if it's just one thing?

**Comment by danoswaltCL** [comment]
Seeing these errors when doing /assign to a group experiment:

{
    "client_session_id": null,
    "endpoint": "/v5/assign",
    "level": "error",
    "message": {
        "className": "IndividualExclusionRepository",
        "code": "42703",
        "details": "Error in assignment",
        "driverError": {
            "code": "42703",
            "file": "parse_relation.c",
            "length": 126,
            "line": "3648",
            "name": "error",
            "position": "282",
            "routine": "errorMissingColumn",
            "severity": "ERROR"
        },
        "file": "parse_relation.c",
        "functionName": "findExcluded",
        "length": 126,
        "line": "3648",
        "parameters": {
            "experimentIds": [
                "83ae5cc5-1e78-4c00-8306-20499058941e",
                "abcd6602-f8c6-4779-adf6-127c16af8cd9",
                "b39586e4-275d-435a-afda-6e9bbbb72ad9",
                "c643eb90-046f-458e-8420-9d9495282ccb"
            ],
            "userId": "locust12"
        },
        "position": "282",
        "query": "SELECT \"individualExclusion\".\"createdAt\" AS \"individualExclusion_createdAt\", \"individualExclusion\".\"updatedAt\" AS \"individualExclusion_updatedAt\", \"individualExclusion\".\"versionNumber\" AS \"individualExclusion_versionNumber\", \"individualExclusion\".\"id\" AS \"individualExclusion_id\", \"individualExclusion\".\"groupId\" AS \"individualExclusion_groupId\", \"individualExclusion\".\"exclusionCode\" AS \"individualExclusion_exclusionCode\", \"individualExclusion\".\"experimentId\" AS \"individualExclusion_experimentId\", \"individualExclusion\".\"userId\" AS \"individualExclusion_userId\", \"experiment\".\"createdAt\" AS \"experiment_createdAt\", \"experiment\".\"updatedAt\" AS \"experiment_updatedAt\", \"experiment\".\"versionNumber\" AS \"experiment_versionNumber\", \"experiment\".\"id\" AS \"experiment_id\", \"experiment\".\"name\" AS \"experiment_name\", \"experiment\".\"description\" AS \"experiment_description\", \"experiment\".\"context\" AS \"experiment_context\", \"experiment\".\"state\" AS \"experiment_state\", \"experiment\".\"startOn\" AS \"experiment_startOn\", \"experiment\".\"consistencyRule\" AS \"experiment_consistencyRule\", \"experiment\".\"assignmentUnit\" AS \"experiment_assignmentUnit\", \"experiment\".\"postExperimentRule\" AS \"experiment_postExperimentRule\", \"experiment\".\"enrollmentCompleteCondition\" AS \"experiment_enrollmentCompleteCondition\", \"experiment\".\"endOn\" AS \"experiment_endOn\", \"experiment\".\"revertTo\" AS \"experiment_revertTo\", \"experiment\".\"tags\" AS \"experiment_tags\", \"experiment\".\"group\" AS \"experiment_group\", \"experiment\".\"conditionOrder\" AS \"experiment_conditionOrder\", \"experiment\".\"assignmentAlgorithm\" AS \"experiment_assignmentAlgorithm\", \"experiment\".\"logging\" AS \"experiment_logging\", \"experiment\".\"filterMode\" AS \"experiment_filterMode\", \"experiment\".\"backendVersion\" AS \"experiment_backendVersion\", \"experiment\".\"type\" AS \"experiment_type\", \"experiment\".\"stratificationFactorStratificationFactorName\" AS \"experiment_stratificationFactorStratificationFactorName\", \"user\".\"createdAt\" AS \"user_createdAt\", \"user\".\"updatedAt\" AS \"user_updatedAt\", \"user\".\"versionNumber\" AS \"user_versionNumber\", \"user\".\"id\" AS \"user_id\", \"user\".\"group\" AS \"user_group\", \"user\".\"workingGroup\" AS \"user_workingGroup\", \"user\".\"originalUserId\" AS \"user_originalUserId\" FROM \"individual_exclusion\" \"individualExclusion\" LEFT JOIN \"experiment\" \"experiment\" ON \"experiment\".\"id\"=\"individualExclusion\".\"experimentId\"  LEFT JOIN \"experiment_user\" \"user\" ON \"user\".\"id\"=\"individualExclusion\".\"userId\" WHERE \"individualExclusion\".\"id\" IN ($1, $2, $3, $4)",
        "routine": "errorMissingColumn",
        "severity": "ERROR",
        "type": "Error in the assignment algorithm"
    },
    "request_id": "8fb2d683-de24-45ba-9d55-5d95fe9fd842",
    "request_method_type": "POST"
}
{
    "client_session_id": null,
    "endpoint": "/v5/assign",
    "level": "error",
    "message": {
        "endPoint": "/api/v5/assign",
        "message": "column individualExclusion.groupId does not exist",
        "name": "QueryFailedError",
        "request": {
            "context": "assign-prog",
            "userId": "locust12"
        },
        "type": "Error in the assignment algorithm"
    },
    "request_id": "8fb2d683-de24-45ba-9d55-5d95fe9fd842",
    "request_method_type": "POST"
}

File: backend/packages/Upgrade/.env.example (hunk 3 above)
Comment (ppratikcr7) [reply backend/packages/Upgrade/.env.example:25]: No, it was part of improvement we found while updating replica connection syntax during this PR. Its kept as an array for consistency with the other replica hostnames var. We can make it a string if you want.

Comment (danoswaltCL) [reply backend/packages/Upgrade/.env.example:25]: ok, this complicates promoting this to production as it's an unexpected configuration change that I don't remember being talked about. @bcb37 @VivekFitkariwala what do you think

File: backend/packages/Upgrade/src/api/controllers/ExperimentClientController.ts (hunk 2 above)
Comment (danoswaltCL) [reply backend/packages/Upgrade/src/api/controllers/ExperimentClientController.ts:1017]: please remove. i have to assume someone was doing this in debugging and accidentally committed it, if this change snuck into prod, anyone with a token would be able to delete our database by doing `DELETE /clearDB`. It makes me nervous even having this endpoint exist in the first place for this exact reason.

**Review by danoswaltCL — CHANGES_REQUESTED** [review]

**Comment by danoswaltCL** [comment]
Can someone please report out exactly what the breaking changes are in this code that we should be aware of? This branch will essentially be promoted directly to staging and then to prod as its own release apart from dev.

There are definitely some commits that exist in 5.1 that I recall being asked to not merge down into dev, please check that also.

**Comment by danoswaltCL** [comment]
Tested the changes in local and in staging with some quick running of locust tests just to get a sense of the API working, looks good no errors, times looked good also.

Here are my notes on things that need changed for deploying to production. Please help me to note other things to check with this update so we are sure prod's version (release/v5.1) is not going to break when we deploy this branch.

notes:
`METRICS` var will need to be modified in `prod`. In `staging` I removed the `METRICS `var in order to get it to run, I have not tested correctly the new format.

[TYPEORM_HOSTNAME_EXPORT_REPLICA](https://github.com/CarnegieLearningWeb/UpGrade/pull/1606/files#diff-d3ff7c7dbc02dbb854a52aef526595e57036c038b4ababd3799b5c1568458baeR25) is a new var, what is the correct usage? Is does this replace the other replica array?

[Please also notice this.](https://github.com/CarnegieLearningWeb/UpGrade/pull/1606/files#diff-295ac941b87696aaedfbf82f419e89fc7217423484b2a7d82f12ad34da811ac8R1096)

@Ben Blanchard we should find an hour or so to swap in the db one more time and test it together

File: backend/packages/Upgrade/src/api/controllers/ExperimentClientController.ts (hunk 2 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/controllers/ExperimentClientController.ts]: code updated

**Comment by ppratikcr7** [comment]
> Tested the changes in local and in staging with some quick running of locust tests just to get a sense of the API working, looks good no errors, times looked good also.
> 
> Here are my notes on things that need changed for deploying to production. Please help me to note other things to check with this update so we are sure prod's version (release/v5.1) is not going to break when we deploy this branch.
> 
> notes: `METRICS` var will need to be modified in `prod`. In `staging` I removed the `METRICS `var in order to get it to run, I have not tested correctly the new format.
> 
> [TYPEORM_HOSTNAME_EXPORT_REPLICA](https://github.com/CarnegieLearningWeb/UpGrade/pull/1606/files#diff-d3ff7c7dbc02dbb854a52aef526595e57036c038b4ababd3799b5c1568458baeR25) is a new var, what is the correct usage? Is does this replace the other replica array?
> 
> [Please also notice this.](https://github.com/CarnegieLearningWeb/UpGrade/pull/1606/files#diff-295ac941b87696aaedfbf82f419e89fc7217423484b2a7d82f12ad34da811ac8R1096)
> 
> @bcb37 we should find an hour or so to swap in the db one more time and test it together

Yeah looks good. Ridham reverted that delete api change. Replica name would be an array.
